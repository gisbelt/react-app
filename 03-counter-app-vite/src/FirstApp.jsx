import React from 'react'
import PropTypes from 'prop-types' //first install prop-types: yarn add prop-types

const getAge = (a,b) =>{
    return a-b;
}

const FirstApp = ( { title, subtitle, name } ) => {
    return (
        <>
        {/* <code>{ JSON.stringify( newMessage ) }</code> */}
        <h1>First App Here</h1>
        <h2 data-testid="test-title">{ title }</h2>
        <p>{ subtitle }</p>
        <p>{ name }</p>
        <h3>My name is: Gisbel Torres</h3>
        <h4>My age is: { getAge(2022,1993) }</h4>
        </>
    )
}

export default FirstApp

FirstApp.propTypes = {
    title: PropTypes.string.isRequired, //title must be string and required
    subtitle: PropTypes.number //subtitle must be a number
}

FirstApp.defaultProps = {
    title: 'There is no title, add one',
    subtitle: 'There is no subtitle',
    name: 'There is no name'
}

// To return several elements is important group them in a parent node
// Like: <></>
// To print an object, we use JSON.stringify() because we can't print objects into {}, but arrays yes
// {title,subtitle} are props and we give them the value in the main.jsx (component FirstApp)