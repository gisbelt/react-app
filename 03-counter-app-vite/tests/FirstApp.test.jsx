// Be sure to have installed React Testing Library: yarn add --dev @testing-library/react
// Also Install yarn add -D jest-environment-jsdom
// Add in jest.config.cjs: 
// module.exports = {
//     testEnvironment: 'jest-environment-jsdom',
// }
// Also install: yarn add -D @babel/preset-react
// Add in babel.config.cjs: 
// [
//     '@babel/preset-react', 
//     {
//         runtime: 'automatic'
//     }
// ],

import { render } from  "@testing-library/react";
import FirstApp from "../src/FirstApp";

describe('Testing component <FirstApp />', () => { 
    // we will take a picture of the component and if anything changes the test will fail.
    test('should match with snapshot', () => { 
        const title = 'Hello, Im Goku!';
        const subtitle = 123;
        //Render component in memory        
        const { container } = render ( <FirstApp title={ title } subtitle={ subtitle } /> )
        //Takes a picture
        expect( container ).toMatchSnapshot();
    })

    //match with text title
    test('should show title in a h2', () => { 
        const title = 'Hello, Im Goku!';
        const subtitle = 123;
        const { container, getByText, getByTestId } = render ( <FirstApp title={ title } subtitle={ subtitle } /> )
        
        expect( getByText(title) ).toBeTruthy();
        expect( getByTestId('test-title').innerHTML ).toBe( title );

        // title must be a h2 and contain a title
        // const h2 = container.querySelector('h2');
        // expect(h2.innerHTML).toContain( title )
    })

    //match with subtitle 
    test('should show subtitle sent by props', () => { 
        const title = 'Hello, Im Goku!';
        const subtitle = 123;
        const { getAllByText } = render ( <FirstApp title={ title } subtitle={ subtitle } /> )
        
        // expects more than one element
        expect( getAllByText(subtitle).length ).toBe(1);
    })
})


// container is an HTML or DOM node
// toMatchSnapshot: takes a picture and creates a folder called _snapshots_ 
// What is the use of that? 
// If tomorrow someone changes a tag of the component, 
// for example an h1 converts it to an h2, the test will fail

// getByText looks for something that matches the text we are sending 
// toBeTruthy: that is 
// toContain: containing a string

// getTestId will help us to take a particular element
// our component have de testid attribute: data-testid="test-title"

// getAllByText return an html elements array, i.e, it expects more than one element