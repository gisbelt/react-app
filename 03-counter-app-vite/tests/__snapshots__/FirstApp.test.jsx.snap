// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing component <FirstApp /> should match with snapshot 1`] = `
<div>
  <h1>
    First App Here
  </h1>
  <h2
    data-testid="test-title"
  >
    Hello, Im Goku!
  </h2>
  <p>
    123
  </p>
  <p>
    There is no name
  </p>
  <h3>
    My name is: Gisbel Torres
  </h3>
  <h4>
    My age is: 
    29
  </h4>
</div>
`;
